src/main/java/org/opensearch/neuralsearch/search/util/HybridSearchSortUtil.java

```diff
@@ -0,0 +1,116 @@
+            }
+        }
+        return new TopFieldDocs[0];
+    }
+
+    private static boolean checkIfTopFieldDocExists(List<TopDocs> topDocs) {
+        // topDocs can be empty if no results found in the shard
+        if (topDocs == null || topDocs.isEmpty()) {
+            return false;
+        }
+        for (TopDocs topDoc : topDocs) {
+            if (topDoc != null && topDoc instanceof TopFieldDocs) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Creates a Sort object based on the provided top field documents.
+     * This method takes an array of top field documents and processes each field to create a corresponding SortField.
+     * The created  SortField objects are then used to create and return a new Sort object.
+     * @param topFieldDocs array of top field documents need to be processed.
+     * @return Sort object created based on provided top field documents.
+     */
+    private static Sort createSort(TopFieldDocs[] topFieldDocs) {
+        if (topFieldDocs == null || topFieldDocs[0] == null) {
+            throw new IllegalArgumentException("topFieldDocs cannot be null when sorting is applied");
+        }
+        final SortField[] firstTopDocFields = topFieldDocs[0].fields;
+        final SortField[] newFields = new SortField[firstTopDocFields.length];
+
+        for (int i = 0; i < firstTopDocFields.length; i++) {
+            final SortField delegate = firstTopDocFields[i];
+            final SortField.Type type = delegate instanceof SortedNumericSortField
+                ? ((SortedNumericSortField) delegate).getNumericType()
+                : delegate.getType();
+
+            if (SortedWiderNumericSortField.isTypeSupported(type) && isSortWideningRequired(topFieldDocs, i)) {
+                newFields[i] = new SortedWiderNumericSortField(delegate.getField(), type, delegate.getReverse());
+            } else {
+                newFields[i] = firstTopDocFields[i];
+            }
+        }
+        return new Sort(newFields);
+    }
+
+    /**
+     * Checks if sort widening is required for the provided top field documents.
+     *
+     * This method iterates through the provided topFieldDocs array and checks if any adjacent pairs of sort fields at the specified index are not equal.
+     * If any such pair is found, it returns true, indicating that sort widening is required. Otherwise, it returns false.
+     * @param topFieldDocs array of top field documents to be checked.
+     * @param sortFieldindex index of the sort field to be checked within each top field document.
+     * @return true if sort widening is required, false otherwise.
+     */
+    private static boolean isSortWideningRequired(TopFieldDocs[] topFieldDocs, int sortFieldindex) {
+        for (int i = 0; i < topFieldDocs.length - 1; i++) {
+            TopFieldDocs currentTopFieldDoc = topFieldDocs[i];
+            TopFieldDocs nextTopFieldDoc = topFieldDocs[i + 1];
```

Comment:
if `topFieldDocs` array is always same array does it mean we're repeating same compute multiple times? Can be save on CPU cycles, maybe compute once

Reply:
No, we need to check for every field because
```
Creates Sort object from topFieldsDocs fields. It is necessary to widen the SortField.Type to maximum byte size for merging sorted docs. Different indices might have different types. This will avoid user to do re-index of data in case of mapping field change for newly indexed data. This will support Int to Long and Float to Double. Earlier widening of type was taken care in IndexNumericFieldData, but since we now want to support sort optimization, we removed type widening there and taking care here during merging.
```